### PR TITLE
EAP7-1144/WFLY-11233 introduce EntityManagerCache

### PR DESCRIPTION
**Pull request by scottmarlow**
https://issues.jboss.org/browse/EAP7-1144
https://issues.jboss.org/browse/WFLY-11233
https://github.com/wildfly/wildfly/pull/11795

### CODE REVIEW

File: jpa/WFLY-11233_EntityManagerCache.adoc [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    The container must throw the IllegalStateException if the application calls EntityManager.close on a container-managed entity manager.
+
+    [87] It is not required that these contracts be used when a third-party persistence provider is not used: the container might use these same APIs or its might use its own internal APIs.
+
+    [88] The container may choose to pool EntityManagers: it instead of creating and closing in each case, it may acquire one from its pool and call clear() on it.
+
+One challenge with introducing an entity manager (per EntityManagerFactory) cache, is knowing when to close EntityManagers after certain errors have occurred on the EntityManager. To deal with this robustness issue, we will discard entity managers that are marked close when retrieving them from the cache. To ensure that we always put the cached EM back in cache, EntityManagerCache use is only done within JTA transactions, since we have control of the release (closing) of the EntityManager and only with SynchronizationType.SYNCHRONIZED (never with SynchronizationType.UNSYNCHRONIZED) persistence contexts.
```
Comment (jamezp) [inline jpa/WFLY-11233_EntityManagerCache.adoc]: This concerns me a bit. This could leave `EntityManager`'s in the cache if one is not retrieved.

**Comment by jamezp** [comment]
@simkam Could you please review this or assign it to someone?

File: jpa/WFLY-11233_EntityManagerCache.adoc (hunk 1 above)
Comment (scottmarlow) [reply jpa/WFLY-11233_EntityManagerCache.adoc]: I agree that caching EntityManager instances that may hold some state past the [EntityManager#clear](https://docs.oracle.com/javaee/7/api/javax/persistence/EntityManager.html#clear--) call (we call this when caching each entity manager).  

An example of an intermittent test failure that [we are seeing now is here](https://gist.github.com/scottmarlow/788a654b9dc53d078093ce64df9d4f26).  We will have to figure out this failure before this feature is complete.  We should review what the hibernate.connection.release_mode setting is, to see if that could be wrong.

Comment (scottmarlow) [reply jpa/WFLY-11233_EntityManagerCache.adoc]: Also seeing another variation of error https://gist.github.com/scottmarlow/9edbe9131fee2a41b3267c78e06f2d0e

Comment (scottmarlow) [reply jpa/WFLY-11233_EntityManagerCache.adoc]: I have yet to prove it but after [WFLY-11360](https://issues.jboss.org/browse/WFLY-11360) is solved, I will test that always running the JPA container Synchronization#afterCompletion call, after the Hibernate ORM Synchronization@afterCompletion call, will resolve the above errors.  I have a local coding change for this that will ensure that we don't put the EntityManager back in the cache, until after Hibernate is done making calls on it (which also includes protecting against the Transaction Manager REAPER thread cancelling transactions in the background).

**Comment by jamezp** [comment]
Closing this as the PR was closed as well.

**Pull request closed without merging**